stdlib/list.gr

```diff
@@ -901,3 +901,1012 @@ provide let sort = (comp, list) => {
+   * all ordered pairs `(a, b)`.
+   *
+   * @param list1: The list to provide values for the first tuple element
+   * @param list2: The list to provide values for the second tuple element
+   * @returns The new list containing all pairs of `(a, b)`
+   *
+   * @since v0.6.0
+   */
+
+  provide let product = (list1, list2) => {
+    fromList(
+      reduceRight((x1, list) => {
+        reduceRight((x2, list) => [(x1, x2), ...list], list, list2)
+      }, [], list1)
+    )
+  }
+
+  /**
+   * Provides the subset of a list given zero-based start index and amount of elements
+   * to include.
+   *
+   * @param start: The index of the list where the subset will begin (inclusive)
+   * @param length: The amount of elements to be included in the subset
+   * @param list: The input list
+   * @returns The subset of the list
+   *
+   * @throws Failure(String): When `start` is negative
+   * @throws Failure(String): When `length` is negative
+   *
+   * @since v0.6.0
+   */
+
+  provide let sub = (start, length, list) => {
+    fromList(sub(start, length, toList(list)))
+  }
+
+  /**
+   * Reverses the elements in the list.
+   * 
+   * @param list: The list to reverse
+   * 
+   * @since v0.6.0
+   */
+
+  provide let reverse = list => {
+    let rec reverseInner = node => {
+      match (node) {
+        None => void,
+        Some(node) => {
+          let next = node.next
+          node.next = node.prev
+          node.prev = next
+          reverseInner(next)
+        },
+      }
+    }
+    reverseInner(list.head)
+    let head = list.head
+    list.head = list.tail
+    list.tail = head
+  }
+
+  /**
+   * Sorts the list.
+   *
+   * Ordering is calculated using a comparator function which takes two array elements and must return 0 if both are equal, a positive number if the first is greater, and a negative number if the first is smaller.
+   * @param comp: The comparator function used to indicate sort order
+   * @param list: The list to be sorted
+   * 
+   * @since v0.6.0
+   */
+
+  provide let sort = (comp, list) => {
+    let { head, tail, length } = fromList(sort(comp, toList(list)))
+    list.head = head
+    list.tail = tail
+    list.length = length
+  }
+
+  /**
+   * Combine the given list of strings into one string with the specified
+   * separator inserted between each item.
+   *
+   * @param separator: The separator to insert between elements
+   * @param list: The list to combine
+   * @returns The combined elements with the separator between each
+   *
+   * @since v0.6.0
+   */
+
+  provide let join = (separator, list) => {
+    // TODO(#728): Improve performance here with buffer approach
+    let iter = (acc, str) => {
+      match (acc) {
+        None => Some(str),
+        Some(prev) => Some(prev ++ separator ++ str),
+      }
+    }
+    match (reduce(iter, None, list)) {
+      None => "",
+      Some(s) => s,
+    }
+  }
+
+  /**
+   * Produces a shallow copy of the input list. The new list contains the
```

Comment:
I dont know if we need to specify it contains the same elements, I feel the terminology shallow copy covers that.